### ISSUE TITLE
Common issues needed across Widgets

### ISSUE THREAD

**Issue by mkarmona** [opentargets/platform#901]
Some of the advance feature we want to support are

- error handling
- graphql field data paginated retrieval
- async waiting data fetching
- 3party api handling

**Error handling**
currently the front end page crashes if there's an error in a widget. 
@mirandaio suggested to look at error boundaries https://reactjs.org/docs/error-boundaries.html

**Pagination**
Currently, tables fetch *all* data rows, with pagination happening in the client.
With API pagination data load will be faster. The table component in `ot-ui` will need to be refactored. We will need to take into account table data **download** and for some widget also any **charts** (e.g. drugs widget)

**3rd party API handling**
Currently the frontend loads through the node.js API data from a number of 3rd party APIs.
In other words: our "demo" API fetches data from external APIs and return it to the frontend as part of the call.
Is this the best approach, or are there other alternatives?
Looking at the code, it appears the API calls the following external APIs:
 * Ensembl - `rest.ensembl.org/homology/id` and `rest.ensembl.org/lookup/id`
 * EPMC - `www.ebi.ac.uk/europepmc/webservices/rest/search`
 * ExpressionAtlas - `https://www.ebi.ac.uk/gxa/json/baseline_experiments`
 * GTEx - `https://www.gtexportal.org/rest/v1/reference/gene` and `https://www.gtexportal.org/rest/v1/expression/geneExpression`
 * PDB - `www.ebi.ac.uk/pdbe/api/mappings/best_structures/`
 * UniProt - `www.ebi.ac.uk/proteins/api/proteins/`

**Comment by LucaFumis** [comment]
I add below details from a Slack conversation with @mkarmona on pagination, to keep for reference.
Some of the details are relevant to specific widgets - I'll try add those in the relevant tickets.

**Question:**
For tables we decided to implement a batch load: the front end loads all rows in chunks (so e.g. for 10000 rows, we might make 2 calls for 5000 rows each), right?

**Answer:** 
please, use 1000 each, the api wont let you ask for 5k at once.

the only widget I would server side paginate is similar diseases and similar targets. that widget has been simplified. it does not contain inner row expansion (unfolding) just the top 20 or 25 , so for that purpose I am returning reversely sorted by default

the rest of the widgets to keep with the same logic doing all on the frontned just batch load them by 1k

one of the widgets we need to follow this logic is knowndrugs by exmaple, as for the plots you need to load everything. We made 2 changes on the backend in order to meet the frontend needs,

1. pre aggregate per row the entry
2. enable batch load as some diseases are quite crowed although far less than before without the pre-aggregation 

trying to be more specific

- **similar targets** - using the count field (to have a maximum of elements noted) just load the first query with page index 0 size 25 and when the user click on next page then load the next 25. The results come already rev sorted by score. Dont allow user to sort by any column, jsut get results and page move
- **known drugs** - here you have to batch load. use the count to see how many pages further you ahve to go to load everything, then with all data you can compute all details of the widget. Please note that you already get some aggregations, uniqueDrugs, uniqueTargets UniqueDiseases to show on the summary. Each row is one (disease, drug, phase, status, target, ...)

**Comment by andrewhercules** [comment]
Ticket closed as `alpha` released
